fairscale/experimental/nn/auto_shard.py

```diff
@@ -0,0 +1,186 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+from functools import reduce
+import logging
+import operator
+from typing import Any, Dict, List
+
+import torch
+import torch.fx
+from torch.fx.node import Node
+
+
+def _get_count(param_count: Dict, node_name: Any) -> int:
+    """Identify different mutations of a given node name."""
+    if node_name in param_count:
+        return param_count[node_name]
+    elif node_name.split("_")[0] in param_count:
+        return param_count[node_name.split("_")[0]]
+    else:
+        raise RuntimeError(f"Unable to find match between param {param_count} and node {node_name}")
+
+
+def _create_shard_to_param_count(param_count: Dict, node_name_to_shard_id: Dict) -> Dict:
+    """Utility to create a map from shard id to param count using existing state."""
+
+    shard_to_param_count: Dict[int, int] = {}
+    for node_name in node_name_to_shard_id.keys():
+        try:
+            count = _get_count(param_count, node_name)
+        except RuntimeError:
+            continue
+        if node_name_to_shard_id[node_name] in shard_to_param_count:
+            shard_to_param_count[node_name_to_shard_id[node_name]] += count
+        else:
+            shard_to_param_count[node_name_to_shard_id[node_name]] = count
+    return shard_to_param_count
+
+
+def _split_nodes(model: Any, shard_count: int = 3) -> Dict:
+    """Utility used to trace a graph and identify shard cutpoints."""
+
+    node_name_to_shard_id: Dict[str, int] = {}
+    shard_id = 0
+    nodes_so_far = []
+    param_count: Dict[str, int] = {}
+    shard_to_param_count = {}
+
+    traced_graph_module = torch.fx.symbolic_trace(model)
+
+    # Find the total number of params in the model and
+    # the number of params per shard we are aiming for.
+    for named_mods in model.named_modules():
+        sum = 0
+        for x in named_mods[1].parameters():
+            mul_dims = reduce(operator.mul, x.size(), 1)
+            sum += mul_dims
+
+        name = named_mods[0].split(".")[0]
```

Comment:
Won't this result in double-counting?

Reply:
Actually it looks like it will be triple counting :) I don't use the individual node params so this does not impact the result. I did modify this to make it more clear.